quinn-udp/benches/throughput.rs

```diff
@@ -1,79 +1,118 @@
+use std::{
+    io::{ErrorKind, IoSliceMut},
+    net::{Ipv4Addr, Ipv6Addr, UdpSocket},
+};
+
 use criterion::{criterion_group, criterion_main, Criterion};
+use tokio::{io::Interest, runtime::Runtime};
+
 use quinn_udp::{RecvMeta, Transmit, UdpSocketState};
-use std::cmp::min;
-use std::net::{Ipv4Addr, Ipv6Addr};
-use std::{io::IoSliceMut, net::UdpSocket, slice};
 
 pub fn criterion_benchmark(c: &mut Criterion) {
     const TOTAL_BYTES: usize = 10 * 1024 * 1024;
-    // Maximum GSO buffer size is 64k.
-    const MAX_BUFFER_SIZE: usize = u16::MAX as usize;
     const SEGMENT_SIZE: usize = 1280;
 
+    let rt = Runtime::new().unwrap();
+    let _guard = rt.enter();
+
     let (send_socket, send_state) = new_socket();
     let (recv_socket, recv_state) = new_socket();
-    // Reverse non-blocking flag set by `UdpSocketState` to make the test non-racy
-    recv_socket.set_nonblocking(false).unwrap();
-
-    let max_segments = min(
-        UdpSocketState::new((&send_socket).into())
-            .unwrap()
-            .max_gso_segments(),
-        MAX_BUFFER_SIZE / SEGMENT_SIZE,
-    );
     let dst_addr = recv_socket.local_addr().unwrap();
 
-    let mut receive_buffer = vec![0; MAX_BUFFER_SIZE];
-    let mut meta = RecvMeta::default();
+    let mut permutations = vec![];
```

Comment:
Instead of this loop, I'd propose turning this into a `const PERMUTATIONS: [(bool, bool); 4]`, which I think will end up being easier to understand. Even after the addition of recvmmsg, I think it's about as compact as the current approach but fully declarative.

(Also seems to me like it would be worthwhile to split the changes to permutations from the commit that makes things async.)

Reply:
@djc are you suggesting something along the lines of:

``` rust
    const PERMUTATIONS: [(bool, bool, bool); 6] = [
        (false, false, false),
        (false, false, true),
        (false, true, false),
        (false, true, true),
        (true, false, true),
        (true, true, false),
        // Windows requires receive buffer to fit entire datagram on GRO
        // enabled socket.
        //
        // OS error: "A message sent on a datagram socket was larger
        // than the internal message buffer or some other network limit,
        // or the buffer used to receive a datagram into was smaller
        // than the datagram itself."
        #[cfg(target_os = "windows")]
        (true, false, false),
        // As above.
        #[cfg(target_os = "windows")]
        (true, true, true),
    ];
```

I don't find this easier to read. It is not obvious to me, that the above covers all permutations of `(bool, bool, bool)`, or contains any duplicates. Nor is it obvious which one would be for `gso_enabled`, `gro_enabled` and `recvmmsg_enabled`.

Reply:
Fair enough, feel free to ignore that.